reference/7.3/Microsoft.PowerShell.Core/About/about_Calling_Generic_Methods.md

```diff
@@ -0,0 +1,69 @@
+---
+description: Describes how to call generic methods of .NET types in PowerShell
+Locale: en-US
+ms.date: 02/01/2022
+schema: 2.0.0
+title: about Calling Generic Methods
+---
+# about_Calling_Generic_Methods
+
+Generics let you tailor a method, class, structure, or interface to the precise data type it acts
+upon. For example, instead of using the <xref:System.Collections.Hashtable> class, which allows keys
+and values to be of any type, you can use the <xref:System.Collections.Generic.Dictionary%602>
+generic class and specify the types allowed for the **key** and **value** properties. Generics
+provide increased code reusability and type safety.
+
+Prior to PowerShell 7.3, you could not use generic methods of a .NET class without complicated
+workarounds using .NET reflection. For an example, see Lee Holmes' blog post
+[Invoking generic methods on non-generic classes in PowerShell](https://www.leeholmes.com/invoking-generic-methods-on-non-generic-classes-in-powershell/).
+
+A generic method is a method with two parameter lists: a list of generic type parameters and a list
+of formal parameters.
+
+The following examples show the new PowerShell syntax for accessing a generic method:
+
+```Syntax
+# static generic methods
+[type_name]::MethodName[generic_type_arguments](method_arguments)
+
+# instance generic methods
+$object.MethodName[generic_type_arguments](method_arguments)
+```
+
+The `generic_type_arguments` can be a a single type, or comma-separated list of types, like
+`[string, int]`, including other generic types like
+`$obj.MethodName[string, System.Collections.Generic.Dictionary[string, int]]()`
+
+For more information, see [Generics in .NET](/dotnet/standard/generics/).
+
+## Example
+
+In this example, we create a list of integers then use the `System.Linq.Enumerable` class to
+enumerate the values and transform them to a new value.
+
+The variable `$list` is a generic `List<T>` object that can only contain integers. `List<T>` is a
+generic class that allows you to specify the type of its members when you create it.
+`[System.Linq.Enumerable]::Select<T,T>()` is a generic method that require two generic type
+parameters and two formal value parameters.
+
+```powershell
+[System.Collections.Generic.List[int]]$list = @( 1, 2, 3, 4, 5 )
+$result = [System.Linq.Enumerable]::Select[int, float](
+    $list,
+    [Func[int, float]]{
+        param($item)
+        [math]::Pow($item, 3)
+    }
+)
+$result
```

Comment:
This example works even without the new generic method feature (see below).
Maybe change it to a code snippet that is enabled by this new feature.

![image](https://user-images.githubusercontent.com/127450/152059081-35f275c0-7db3-4f02-8856-d34999cbae6d.png)

Reply:
This was one of the pester tests. I tried this in 7.3-preview.1 it does not work.

```
ParserError:
Line |
   1 |  $result = [System.Linq.Enumerable]::Select[int, float](
     |                                             ~
     | Array index expression is missing or not valid.
```

I tested in a current build of master. It works there.

Reply:
Without the `[int, float]` part. Please take a closer look at the screenshot above :)

Reply:
OK, I see that now. The example, as written, is enabled by the new feature. But, as you point out, you can remove types and it works. Can anyone provide a better example?

cc: @vexx32

Reply:
Currently the main things that are strictly _enabled_ by this are methods with zero arguments for PS to infer from, for example:

```powershell
[System.Linq.Enumerable]::Empty[string]()
[array]::Empty[int]()
```

While it's not yet the case, it could also be used to improve generic method resolution to require less casts (in theory). For example, currently to call `[System.Linq.Enumerable]::Select()` you have to cast some or all of the arguments to specific types as you have above.

For example, it should be possible to reduce the call above to `[System.Linq.Enumerable]::Select[int, float]($list, { <# script block stuff #> })` and avoid the additional casts which are currently necessary.

I think the example as written is fine to illustrate the usage, even if it's not strictly required in all cases currently. Some folks may well just use it to be explicit about exactly what they're calling.

Method resolution can be very complicated, and there can be cases where a method may have both generic and non-generic overloads, where PS could have trouble resolving or may resolve incorrectly (from the user's perspective) without the explicit generic method arguments.

Reply:
Without a better example, I think this illustrates the point (@vexx32 said).

Reply:
> For example, it should be possible to reduce the call above to [System.Linq.Enumerable]::Select[int, float]($list, { <# script block stuff #> }) and avoid the additional casts which are currently necessary.

This doesn't work for `Select`, because there is still ambiguity in the type of the delegate as an argument. See the screenshot below:

![image](https://user-images.githubusercontent.com/127450/152206022-10a85c17-84cb-4db0-b40a-1c378c844571.png)

But it works for `IEnumerable.Any` -- when specifying the generic type argument (`[int]` in this case), no need to have explicit cast for the script block.

![image](https://user-images.githubusercontent.com/127450/152208820-a1f4217e-8a1a-4573-9a24-405a7a17b2b0.png)

-----------

I'm fine to keep the current example, or use the `Any` method as an example (it's up to you @sdwheeler). But please add another one for `[array]::Empty[int]()`, because that's the scenario enabled by this new feature, which doesn't work before.

Reply:
Ok. Will add `[Array]::Empty[int]()`.